libraries/chain/include/eosio/chain/apply_context.hpp

```diff
@@ -621,6 +628,8 @@ class apply_context {
       bool                          privileged   = false;
       bool                          context_free = false;
 
+      std::optional<sync_call_context> sync_call_ctx{};  // only one of act and sync_call_ctx can be present
```

Comment:
Instead of having `apply_context` having members that can be there or not, I think it would be cleaner to have something like (also that would remove the need for a separate class `sync_call_context`):

```
struct context_base {
   ...  // all common members
};

struct apply_context : public context_base {
   uint32_t recurse_depth; ///< how deep inline actions can recurse
   uint32_t first_receiver_action_ordinal = 0;
   uint32_t action_ordinal                = 0;
};

struct call_context : public context_base {
   account_name          sender;
   account_name          receiver;
   std::span<const char> data; // includes function name, arguments, and other information
};
```

Reply:
Thanks. The Implementation section of the design document talked about this too. This PR focuses on demonstrating the feasibility of sync calls. I will create a task to track it.

Reply:
https://github.com/AntelopeIO/spring/issues/1261